src/EditorFeatures/Core/Snippets/RoslynLSPSnippetExpander.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Reflection;
+using System.Text;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+
+namespace Microsoft.CodeAnalysis.Snippets
+{
+    [Export(typeof(IRoslynLSPSnippetExpander))]
+    [Export(typeof(RoslynLSPSnippetExpander))]
+    internal class RoslynLSPSnippetExpander : IRoslynLSPSnippetExpander
+    {
+        protected object? _lspSnippetExpander;
+        protected Type? _expanderType;
+        protected MethodInfo? _expanderMethodInfo;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RoslynLSPSnippetExpander(
+            [Import("Microsoft.VisualStudio.LanguageServer.Client.Snippets.LanguageServerSnippetExpander", AllowDefault = true)] object? languageServerSnippetExpander)
+        {
+            _lspSnippetExpander = languageServerSnippetExpander;
+
+            if (_lspSnippetExpander is not null)
+            {
+                _expanderType = _lspSnippetExpander.GetType();
+                _expanderMethodInfo = _expanderType.GetMethod("TryExpand");
```

Comment:
It already is public in 17.3-Preview1 BUT because we're trying to ensure that public consumers of dotnet/roslyn can use public previews it's utilizing reflection to ensure those folks can continue to build the repo